feat(scoped-storage): Add consent dialog

david-allison-1 opened the pull request:
To release on the Play Store, we need to implement scoped storage

Since the migration to scoped storage can take a while, and has a low,
but nonzero chance of data corruption, we want to show a dialog to the
user, ensuring that:

* They understand that we need to do this
* They understand the risks
* They have backed up
* They won't uninstall the app before/during
  the migration

We also add a link to https://github.com/ankidroid/Anki-Android/wiki/Storage-Migration-FAQ
to ensure that users are informed.

Since we can set "preserveLegacyExternalStorage", we don't need the
user to immediately migrate, but this stops working if the user
uninstalls, and the directory becomes inaccessible

Issue #5304

## How Has This Been Tested?

![image](https://user-images.githubusercontent.com/62114487/143665690-d98ce326-09f2-4215-8630-ecaf23a16ce6.png)
![image](https://user-images.githubusercontent.com/62114487/143665692-a895f2e9-c61a-4253-937d-b8ee4b637dcb.png)
![image](https://user-images.githubusercontent.com/62114487/143665695-777fef8e-8b88-4879-ad43-04a0537e48be.png)


## Learning (optional, can help others)
[Please see /docs/ changes](https://github.com/david-allison-1/Anki-Android/tree/b60893bf5c73fc089a3aaed01d0587af3131f38d/docs/scoped_storage)

https://github.com/ankidroid/Anki-Android/wiki/Storage-Migration-FAQ

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

david-allison-1 commented:
This is for design review.

My main issue: This doesn't inform the user that they will be unable to sync

I think we should move the "time taken" to the second screen

mikehardy commented on AnkiDroid/src/main/res/values/02-strings.xml:
```diff
@@ -408,4 +408,16 @@
 
+    <!-- Scoped Storage Migration -->
+    <string name="scoped_storage_title">Storage migration</string>
+    <string name="scoped_storage_initial_message">AnkiDroid needs to move its data due to Android privacy rules
+\n\nEstimated duration: %s
+\n\nDo not uninstall AnkiDroid before migrating data</string>
```
```suggestion
    <string name="scoped_storage_initial_message">AnkiDroid must move its data to meet new Android privacy rules
\n\nIf you uninstall AnkiDroid before the migration, your data will be lost.
\n\nPlease perform the migration as soon as possible.</string>
```

mikehardy commented:
Some light word-smithery, based on your idea of moving the estimate to a following page, I agree with that.
I like the way this is going and in general think "yes, go for it"

david-allison-1 commented:
Screenshots updated. Went with slight wording changes, 

* added text to the second screen
* Changed second screen color to match (surprisingly difficult)

mikehardy reviewed (approved):
Fantastic, merge at will!

david-allison-1 commented:
I'll ping @Arthur-Milchior as this is non-blocking, but happy to merge

voidplayer commented:
> The problem is, we really want people to migrate - if they uninstall the app, they permanently lose access to their AnkiDroid folder, and would need a PC + USB (or AnkiWeb) to recover their data.

They could also reinstall anki to recover access to their data, couldnt they? That sounds like a good way to deal with this

I was going to comment a few other things, but i think it looks really good (user experince wise)

david-allison-1 commented:
> They could also reinstall anki to recover access to their data, couldnt they? That sounds like a good way to deal with this

No, AnkiDroid permanently loses access to storage on their phone except specific private directories once they uninstall+reinstall on Android 11 and above (thanks Google):

Ctrl+F for `preserveLegacyExternalStorage`: https://developer.android.com/training/data-storage/use-cases

These private directories are deleted when a user uninstalls the app (but we can add an OS provided "are you sure you want to delete your data").

voidplayer commented:
I see

What i meant is that anki dont lose access in this upgrade. Something goes terribly wrong, it can still recover until the version where it cant

voidplayer commented:
And by that i mean that user can report a bug, and be fixed in 2.16.1 because you still have access

mikehardy commented:
Going to go with a merge here so we can keep on moving

Arthur-Milchior commented:
Sorry for having been absent so long.

I'm really confused.
We actually know whether the user has data on their device that was not synced. Why do we ask ?  
They can have data on ankiweb that is not sync, but it does not matter.

Note that I do not want it to be mandatory. I even suspect we would be in breach of our declaration to Android play store, as we mentioned that user do not have to send us any data, and with this feature, data transfer would be virtually mandatory, or at least we will be really insistant about it. Plus, users may well have sensitive data that they would not trust ankiweb to host. Its a free service and while AnkiTekts terms state data is protected, the actual details are absent about what is actually done to ensure protection.


I love the recommendation not to move the data while the app is being moved.

One thing is not clear right now, if the user decide to postpone, how can they tell ankidroid to migrate. Even if we keep bugging the user until they accept, they need to know it in advance, so that they can postpone without fear.

I would love "learn more" to be on all page. After all, if they click "migrate" and then they can't learn more, that seems a problem; as they may agree because they trust us and then want to understand.



I'm really not a fan of having important information online. Not only due to internet price, but because we have an awful track record and no tool of documentation translation and updating. Using crowdin would at least ensure we get a way to let translators know what need to be translated

Arthur-Milchior commented:
Sorry for having been absent so long.

I'm really confused.
We actually know whether the user has data on their device that was not synced. Why do we ask ?  
They can have data on ankiweb that is not sync, but it does not matter.

Note that I do not want it to be mandatory. I even suspect we would be in breach of our declaration to Android play store, as we mentioned that user do not have to send us any data, and with this feature, data transfer would be virtually mandatory, or at least we will be really insistant about it. Plus, users may well have sensitive data that they would not trust ankiweb to host. Its a free service and while AnkiTekts terms state data is protected, the actual details are absent about what is actually done to ensure protection.


I love the recommendation not to move the data while the app is being moved.

One thing is not clear right now, if the user decide to postpone, how can they tell ankidroid to migrate. Even if we keep bugging the user until they accept, they need to know it in advance, so that they can postpone without fear.

I would love "learn more" to be on all page. After all, if they click "migrate" and then they can't learn more, that seems a problem; as they may agree because they trust us and then want to understand.



I'm really not a fan of having important information online. Not only due to internet price, but because we have an awful track record and no tool of documentation translation and updating. Using crowdin would at least ensure we get a way to let translators know what need to be translated

david-allison commented:
> We actually know whether the user has data on their device that was not synced. Why do we ask ?
> They can have data on ankiweb that is not sync, but it does not matter.

Agreed. We can easily fix this

> Note that I do not want it to be mandatory...

Absolutely, the screen shows "local backup" if the user doesn't have an AnkiWeb account. 

We should add "local backup" in the rare case that the user has AnkiWeb, but isn't using it. 

I feel we want an option to "full local backup", rather than informing the user to do so

> One thing is not clear right now, if the user decide to postpone, how can they tell ankidroid to migrate. Even if we keep bugging the user until they accept, they need to know it in advance, so that they can postpone without fear.

Right now, this is TBC and open to discussion

> I would love "learn more" to be on all page. After all, if they click "migrate" and then they can't learn more, that seems a problem; as they may agree because they trust us and then want to understand.

Absolutely!

> I'm really not a fan of having important information online. Not only due to internet price, but because we have an awful track record and no tool of documentation translation and updating. Using crowdin would at least ensure we get a way to let translators know what need to be translated

Unsure: Using CrowdIn means the user needs to update the app to get the fixes

mikehardy commented:
> > I'm really not a fan of having important information online. Not only due to internet price, but because we have an awful track record and no tool of documentation translation and updating. Using crowdin would at least ensure we get a way to let translators know what need to be translated

> Unsure: Using CrowdIn means the user needs to update the app to get the fixes

This part in particular I differ on. I *prefer* items like this to be on a web page. 

regarding price+availability: I believe we can make a general assumption that users have internet some times at least. If we allow them to postpone, so they can choose their time when they have a comfortable internet connection I think this is okay, and I live in the developing world - I'm keenly aware of people that have no data connections on their phone.

regarding translation 1: our translators themselves are slow, and incomplete even if the text is complete+correct. Stated more strongly: relying on translation here is actually an error in my mind. No matter how much lead time we try to achieve, *most* languages will not have translations. If we use a web page at least users can use the widely available translation tools the world has created

regarding translation 2: we will unavoidably learn lots of things users need to know or care about after release, and with a web page we can update those out of cycle with respect to app release. Editing a wiki page let's users (or us) easily help all users immediately, instead of doing an app release that may be blocked for some reason in app review (it happens!).

Arthur-Milchior commented:
Okay, let's go with web page

Akshay0701 commented:
not going to comment on the content of the dialog, because I am not a native English speaker,

regrading design of dialog, David in the first image you set all buttons in single vertical order, first of all its not common in all apps, having a horizontal button with most important one on right side, less important on left side as below will make things clear to users:

![Capture](https://user-images.githubusercontent.com/52353967/162628928-0f3c222c-cc3a-45e2-a574-25284c720d77.png)

and if you want every user to click on "migrate" button then give green color to that button, and red color to postpone button, this design will make users known that they should migrate.

david-allison commented:
That's the default for our library: I didn't specifically design the vertical options.

Absolutely agreed on the color

BrayanDSO commented:
I agree with the `Red` and `Green` button color

Suggestion 1: disable the Migrate option if the checkboxes aren't checked instead of a toast

Suggestion 2: Add a toolbar button to upgrade if the user chooses `Postpone` with the same alert of sync.

![WhatsApp Image 2022-04-10 at 3 02](https://user-images.githubusercontent.com/69634269/162633393-2bad8307-d801-4a93-8e65-72743c03a972.png)

Another option would be something like a persistent snackbar in the deck picker activity.

Either way, I think every week or any discussed time period there should be a prompt to migrate

mikehardy commented:
I like all of those suggestions @BrayanDSO - best part about this bit is that it can go in parallel to David's work. If someone wants to peel out a separate issue (making sure there isn't one first) for this, then it can be available for dev

UvrajSB commented:
Why we are asking whether the user has currently synced the data or not, cant we sync it right before the start of the migration process regardless of when he synced

UvrajSB commented:
I have come up with this flow, please suggest changes, and pardon me if my handwriting is not legible somewhere
![WhatsApp Image 2022-04-11 at 4 48 45 AM](https://user-images.githubusercontent.com/73571511/162644239-e660b569-ecee-4afa-a8ab-96357ae5f127.jpeg)

UvrajSB commented:
I am considering @BrayanDSO 's suggestion #2 in my model.
I think we should remove the "never ask again" option.

Arthur-Milchior commented:
In any case, I believe that this should be in another issue, because most of us probably don't look at text in merged pull request
